subprojects/logging/src/main/java/org/gradle/internal/logging/sink/ProgressLogEventGenerator.java

```diff
@@ -45,169 +50,174 @@
+        lastRenderedBuildOpId = null;
+        listener.onOutput(event);
+    }
 
-        if (!deferHeader || !(progressStartEvent.getLoggingHeader() != null && progressStartEvent.getLoggingHeader().equals(progressStartEvent.getShortDescription()))) {
-            operation.startHeader();
+    // Format consistent with 3.x and lower versions of Gradle to allow automated parsing (not officially supported)
+    private LogEvent plainHeader(ProgressStartEvent startEvent) {
+        if (startEvent.getLoggingHeader() != null) {
+            return new LogEvent(startEvent.getTimestamp(), startEvent.getCategory(), startEvent.getLogLevel(), startEvent.getLoggingHeader(), null, startEvent.getBuildOperationId());
         }
+        return null;
     }
 
-    enum State {None, HeaderStarted, HeaderCompleted, Completed}
+    // Return the id of the operation/group, checking up the build operation hierarchy
+    private Object getOperationId(@Nullable final Object buildOpId) {
+        Object current = buildOpId;
+        while (current != null) {
+            if (operationsInProgress.containsKey(current)) {
+                return current;
+            }
+            current = buildOpIdHierarchy.get(current);
+        }
+        return null;
+    }
 
     private class Operation {
-        private final Object buildOperationIdentifier;
+        private final Object buildOpIdentifier;
         private final String category;
         private final String loggingHeader;
         private final long startTime;
-        private final boolean hasLoggingHeader;
-        private String status = "";
-        private State state = State.None;
-        private long completeTime;
 
-        private Operation(String category, String loggingHeader, long startTime, Object buildOperationIdentifier) {
+        private List<RenderableOutputEvent> bufferedLogs = new ArrayList<RenderableOutputEvent>();
+        private String status;
+
+        private Operation(String category, @Nullable String loggingHeader, long startTime, @Nullable Object buildOpIdentifier) {
             this.category = category;
             this.loggingHeader = loggingHeader;
             this.startTime = startTime;
-            this.hasLoggingHeader = GUtil.isTrue(loggingHeader);
-            this.buildOperationIdentifier = buildOperationIdentifier;
+            this.buildOpIdentifier = buildOpIdentifier;
         }
 
-        private StyledTextOutputEvent plainTextEvent(long timestamp, String text) {
-            return new StyledTextOutputEvent(timestamp, category, LogLevel.LIFECYCLE, buildOperationIdentifier, Collections.singletonList(new StyledTextOutputEvent.Span(text)));
+        StyledTextOutputEvent header(final String message) {
+            StyledTextOutputEvent.Span span = new StyledTextOutputEvent.Span(StyledTextOutput.Style.Header, "> " + message + EOL);
+            return new StyledTextOutputEvent(startTime, category, LogLevel.QUIET, buildOpIdentifier, Collections.singletonList(span));
```

Comment:
Should the log level really be `QUIET`? That would mean you could never turn it off even if you provide `-q`. From my perspective the intension of quiet is to show almost nothing.